File: packages/jexlang-swift/jexlang-swift/eval/transforms/MapTransformRegistry.swift
//
//  MapTransformRegistry.swift
//  jexlang-swift
//
//  Created by Vinay Kumar on 04/11/25.
//

import Foundation

public final class MapTransformRegistry: TransformRegistry {
    private var map: [String: TransformImpl]

    public init(_ transformMap: [String: TransformImpl] = [:]) {
        self.map = transformMap
    }

    public func set(_ name: String, _ fn: @escaping TransformImpl) {
        map[name] = fn
    }

    public func getAll() -> [String: TransformImpl] {
        map
    }

    public func has(_ name: String) -> Bool {
        map[name] != nil
    }

    public func remove(_ name: String) {
        map.removeValue(forKey: name)
    }

    public func transform(_ name: String, _ input: JexValue, _ ctx: EvaluatorContext) -> JexValue {
        guard let fn = map[name] else {
            fatalError("Unknown transform: \(name)")
        }
        return fn(input, ctx)
    }
}

File: packages/jexlang-java/src/main/java/io/github/vinaykumar0339/transforms/MapTransformRegistry.java
package io.github.vinaykumar0339.transforms;

import io.github.vinaykumar0339.context.EvaluatorContext;
import io.github.vinaykumar0339.types.JexValue;

import java.util.Map;

public final class MapTransformRegistry implements TransformRegistry {
    private final Map<String, TransformImpl> map;

    public MapTransformRegistry(Map<String, TransformImpl> transformMap) {
        this.map = transformMap;
    }

    public void set(String name, TransformImpl fn) { map.put(name, fn); }
    public Map<String, TransformImpl> getAll() { return map; }
    public boolean has(String name) { return map.containsKey(name); }
    public void remove(String name) { map.remove(name); }
    public JexValue transform(String name, JexValue input, EvaluatorContext ctx) {
        TransformImpl f = map.get(name);
        if (f == null) throw new RuntimeException("Unknown transform: " + name);
        return f.apply(input, ctx);
    }
}
